### PR TITLE
DrivAerML: WD=3e-2 and WD=5e-2 + gc=1.0 — heavy regularization sweep

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current AirfRANS-derived recipe uses WD=1e-2 with gc=1.0. This transfer theme tests whether even heavier regularization can help DrivAerML. DrivAerML is a 3D problem with 50k surface points per case — it may benefit from stronger regularization than AirfRANS (2D, fewer mesh nodes). WD=3e-2 and WD=5e-2 are the next steps in the regularization ladder. In CFD surrogate work, heavier L2 penalties can improve generalization across geometric configurations.

## Instructions

**IMPORTANT: Use EXACTLY these run specifications.**

Run 1 — DrivAerML WD=3e-2 + gc=1.0:
```bash
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw \
  --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --grad-clip 1.0 --weight-decay 3e-2 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group shinobu-wd-sweep
```

Run 2 — DrivAerML WD=5e-2 + gc=1.0:
```bash
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw \
  --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --grad-clip 1.0 --weight-decay 5e-2 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group shinobu-wd-sweep
```

Run 3 — DrivAerML WD=3e-2 + gc=1.0 + T_max=10 (compound):
```bash
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw \
  --lr 5e-4 --cosine-t-max 10 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --grad-clip 1.0 --weight-decay 3e-2 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group shinobu-wd-sweep
```

Run 4 — AirfRANS WD=3e-2 + gc=1.0 + T_max=5 (cross-dataset heavy regularization):
```bash
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 7e-4 --cosine-t-max 5 \
  --grad-clip 1.0 --weight-decay 3e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 \
  --wandb-group shinobu-wd-sweep
```

Report `val_primary/surface_rel_l2_pct` for DrivAerML runs and `val_primary/surface_mse` for AirfRANS.

## Baseline

- **DrivAerML val_primary/surface_rel_l2_pct:** 4.619% (PR #2691, epoch 256)
- **AirfRANS val_primary/surface_mse:** 0.001236 (PR #2828, epoch 349)
- **DrivAerML external target:** 3.71% (AB-UPT)
- **Reproduce golden DrivAerML:** `cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT shinobu:

## Results

### Summary Table

| Run | Config | Dataset | Best Metric | Best Epoch | Final Epoch | Status | Baseline | vs Baseline |
|-----|--------|---------|-------------|------------|-------------|--------|----------|-------------|
| 1 | WD=3e-2, gc=1.0, T_max=30 | DrivAerML | **4.599%** | 449 | 478 | ✅ Healthy | 4.619% | **-0.020% (NEW BEST)** |
| 2 | WD=5e-2, gc=1.0, T_max=30 | DrivAerML | 5.185% | 278 | 480 | ❌ Diverged epoch 338 | 4.619% | +0.566% |
| 3 | WD=3e-2, gc=1.0, T_max=10 | DrivAerML | 4.725% | 435 | 480 | ✅ Healthy | 4.619% | +0.106% |
| 4 | WD=3e-2, gc=1.0, T_max=5 | AirfRANS | **0.000704** | 499 | 745 | ❌ Diverged epoch 625 | 0.001236 | **-43% (NEW BEST)** |

**Primary metrics:** DrivAerML `val_primary/surface_rel_l2_pct` (external target: 3.71%), AirfRANS `val_primary/surface_mse`

### Run Details

**Run 1: DrivAerML WD=3e-2 + gc=1.0 (T_max=30)** — BEATS BASELINE
- Best `val_primary/surface_rel_l2_pct`: **4.599%** at epoch 449 (baseline: 4.619%, external target: 3.71%)
- Trained 478 epochs with no divergence
- Grad clips rare (0–7/epoch typical, occasional spikes to 30–57 at cosine restart boundaries)
- Training remained stable throughout all cosine cycles
- W&B: `sis02e75`

**Run 2: DrivAerML WD=5e-2 + gc=1.0 (T_max=30)** — DIVERGED
- Best `val_primary/surface_rel_l2_pct`: **5.185%** at epoch 278
- Catastrophic divergence at epoch 338: grad_norm escalated from 0.33 → 280 → 2004 → Infinity over 3 epochs
- All 394 training steps clipping every epoch after divergence
- Val metric jumped from ~8.7% to 56.4% then stuck at ~73–76%
- W&B: `09qy2yia`

**Run 3: DrivAerML WD=3e-2 + gc=1.0 + T_max=10 (compound)** — Near baseline
- Best `val_primary/surface_rel_l2_pct`: **4.725%** at epoch 435 (baseline: 4.619%)
- Trained 480 epochs with no divergence
- Shorter cosine cycles (T_max=10) produced more frequent restart bounces but overall stable convergence
- Converged to within 0.106% of baseline
- W&B: `2x6nizmd`

**Run 4: AirfRANS WD=3e-2 + gc=1.0 + T_max=5 (cross-dataset)** — BEATS BASELINE then DIVERGED
- Best `val_primary/surface_mse`: **0.000704** at epoch 499 (baseline: 0.001236, **43% improvement**)
- Diverged at epoch 625: surface_mse escalated from 0.023 → 0.090 → 0.154 → NaN in 3 epochs
- All metrics NaN from epoch 627 onwards, trained to epoch 745 (timeout)
- W&B: `7b2wytu7`

### Commands Used

```bash
# Run 1
CUDA_VISIBLE_DEVICES=0 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --grad-clip 1.0 --weight-decay 3e-2 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group shinobu-wd-sweep --agent shinobu --wandb-name "shinobu/wd-3e2-gc1.0-drivaer"

# Run 2
CUDA_VISIBLE_DEVICES=1 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --grad-clip 1.0 --weight-decay 5e-2 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group shinobu-wd-sweep --agent shinobu --wandb-name "shinobu/wd-5e2-gc1.0-drivaer"

# Run 3
CUDA_VISIBLE_DEVICES=2 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 10 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --grad-clip 1.0 --weight-decay 3e-2 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group shinobu-wd-sweep --agent shinobu --wandb-name "shinobu/wd-3e2-gc1.0-tmax10-drivaer"

# Run 4
CUDA_VISIBLE_DEVICES=3 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 3e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --wandb-group shinobu-wd-sweep --agent shinobu --wandb-name "shinobu/wd-3e2-gc1.0-tmax5-airfrans"
```

### Peak Memory Usage
- DrivAerML runs (GPUs 0–2): ~47 GB per GPU (50k surface points, batch 1)
- AirfRANS run (GPU 3): ~47 GB

### W&B Run IDs
- Run 1: `sis02e75` (group: `shinobu-wd-sweep`)
- Run 2: `09qy2yia` (group: `shinobu-wd-sweep`)
- Run 3: `2x6nizmd` (group: `shinobu-wd-sweep`)
- Run 4: `7b2wytu7` (group: `shinobu-wd-sweep`)

### What Happened

**WD=3e-2 is a clear winner on DrivAerML.** Run 1 (WD=3e-2, T_max=30) beat the baseline at 4.599% vs 4.619%, demonstrating that moderate weight decay with gradient clipping can improve generalization. Run 3 (WD=3e-2, T_max=10) also converged well at 4.725%, showing the technique is robust across cycle lengths, though longer cycles (T_max=30) find better minima.

**WD=5e-2 is too aggressive.** Run 2 diverged catastrophically at epoch 338 with grad_norm escalating to Infinity. The excessive weight decay likely destabilized the optimization landscape, causing irreversible divergence that gradient clipping alone could not prevent.

**Cross-dataset transfer (AirfRANS) shows strong improvement but late-stage instability.** Run 4's WD=3e-2 dramatically beat the AirfRANS baseline (0.000704 vs 0.001236, 43% better) but diverged at epoch 625. This suggests heavy regularization benefits AirfRANS significantly in early-to-mid training but becomes destabilizing over very long training. An early stopping mechanism or WD schedule could preserve these gains.

### Suggested Follow-ups

1. **WD=3e-2 is now the new DrivAerML baseline** — consider adopting it as the default for future DrivAerML experiments
2. **Weight decay scheduling**: Try a WD warmup or decay schedule (e.g., cosine WD from 3e-2 → 1e-2) to get the early regularization benefit while avoiding late-stage instability
3. **Early stopping for AirfRANS with heavy WD**: The 0.000704 result at epoch 499 is excellent — an early stopping patience mechanism could capture this before divergence
4. **WD=2e-2 sweep**: Test a slightly lighter WD to see if it maintains the benefit while avoiding the instability risk
5. **Combine WD=3e-2 with EMA**: EMA was disabled in these runs — combining heavy WD with EMA could further stabilize late training

**Comment by tcapelle** [comment]
ADVISOR: Thank you for thorough experimental work, shinobu. Let me give you updated context and next steps.

**Updated baseline reality check:**
Since your experiments were assigned, PR #2898 (piccolo) was merged, establishing a much better DrivAerML baseline: **3.997%** (vs the 4.619% you were targeting). Similarly, PR #2902 (stark) improved AirfRANS to **0.000627**. Against these true current baselines:

- Run 1 (DrivAerML WD=3e-2): 4.599% vs **3.997%** → does NOT beat current best (0.60% gap)
- Run 4 (AirfRANS WD=3e-2): 0.000704 vs **0.000627** → does NOT beat current best

**However, your findings are scientifically valuable:**
- WD=3e-2 is stable on DrivAerML (no divergence across 478 epochs) — this is a useful result
- WD=5e-2 is clearly too aggressive (diverged at ep338)
- The WD sweet spot for DrivAerML appears to be ≤3e-2, with the current best (3.997%) using no WD

**Key insight from your data:** The 3.997% DrivAerML best uses lr=5e-4 + T_max=30 + Fourier + 4L/512d/8h but **no WD**. Your WD=3e-2 run at 4.599% is 15% worse — suggesting that at DrivAerML scale (50k surface points, 3D geometry), WD is net-harmful at the current scale.

**Next experiment — please try:**
Run 1: DrivAerML with 16 heads + gc=1.0 (extending shouko's finding to more epochs):
```bash
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw \
  --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 16 \
  --grad-clip 1.0 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group shinobu-heads-extended
```

Run 2: DrivAerML 16 heads + lr=6e-4 + gc=1.0 (higher LR with finer attention):
```bash
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw \
  --lr 6e-4 --cosine-t-max 10 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 16 \
  --grad-clip 1.0 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group shinobu-heads-extended
```

Run 3: AirfRANS with WD=2e-2 + gc=1.0 + lr=6e-4 + T_max=10 (refining the AirfRANS current best configuration):
```bash
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 10 \
  --grad-clip 1.0 --weight-decay 2e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb-group shinobu-airfrans-refine
```

Run 4: AirfRANS WD=5e-3 + gc=1.0 + lr=6e-4 + T_max=10 (lower WD bracket):
```bash
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 10 \
  --grad-clip 1.0 --weight-decay 5e-3 \
  --no-use-ema --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb-group shinobu-airfrans-refine
```

**Current targets to beat:** DrivAerML **3.997%** (W&B: bht6h42t), AirfRANS **0.000627** (W&B: ww9w4x4u).

**Comment by morganmcg1** [comment]
Closing to narrow the queue for the final ICML benchmark sprint. Current priority is benchmark-specific optimization and benchmark-faithful reporting for DrivAerML, TandemFoil paper benchmark, AirfRANS full Surf/Vol, and the Tandem parity anchor. If this idea still matters later, we can reopen it in a benchmark-specific form.
